### DISCUSSION TITLE
[8 febbraio 2024] Complessità

### DISCUSSION THREAD

**Discussion by Antaarees** [sapienzastudentsnetwork/automi-calcolabilita-complessita#22] [Complessità]
![image](https://github.com/sapienzastudentsnetwork/acc2324/assets/69389461/bda8c73b-a6df-4feb-8c80-562dd141a40e)

**Comment by Exyss** [comment]
Per il primo punto, ripropongo la soluzione presente sui miei appunti:

![image](https://github.com/user-attachments/assets/8fa01180-da0d-4284-97f5-492c3f916111)

----

La dimostrazione richiesta nel secondo punto è presente sul libro di testo. La seconda parte della dimostrazione invece abbiamo che:

![image](https://github.com/user-attachments/assets/5d004d8e-8e5d-4072-871f-2e7efe5df0fb)

**Comment by Elia-Belli** [comment]
**Teorema**

$$4COL\leq_m^p SAT$$

**Dimostrazione**
Per il _Teorema di Cook-Levin_ sappiamo che $SAT\in NP-Complete$ , quindi $\forall L\in NP,\ L\leq_m^p SAT$ .

Di conseguenza basta mostrare che $4COL\in NP$, ad esempio descrivendo un verificatore polinomiale $V$ su input $<G,y>$:
1. Interpreta $y=(c_1,c_2,\dots,c_n)$ come una 4-colorazione per $G$, con $c_i\in \\{C,M,Y,K\\}$, se non è una 4-colorazione rifiuta
2. Verifica $y$ sia una 4-colorazione valida: $\forall (i,j)\in E(G)$ verifica $c_i\neq c_j$, altrimenti rifiuta
3. Accetta

La complessità di tempo è ovviamente polinomiale $\Rightarrow V$ è un verificatore polinomiale per $4COL\ \ \square$
